spacy/tokens/doc.pyx

```diff
@@ -1743,6 +1743,124 @@ cdef class Doc:
+        ps_search_chars: a byte array containing, in numerical order, UTF-8 characters to search for within each token, 
+            starting at the beginning.
+        ps_width_offsets: an array of single-byte values [1-char-start, 2-char-start, 3-char-start, 4-char-start, 4-char-end]
+            specifying the offsets within *ps_search_chars* that contain UTF-8 characters with the specified widths.
+        ps_lengths: an array of single-byte values specifying the lengths of search results (from the beginning) to be hashed 
+            in ascending order. For example, if *ps_lengths==[1, 2]*, *ps_search=="aC" and *cs==False*, the searched strings 
+            hashed for "spaCy" would be "a" and "ac".
+        ss_search_chars: a byte array containing, in numerical order, UTF-8 characters to search for within each token, 
+            starting at the end.
+        ss_width_offsets: an array of single-byte values [1-char-start, 2-char-start, 3-char-start, 4-char-start, 4-char-end]
+            specifying the offsets within *ss_search_chars* that contain UTF-8 characters with the specified widths.
+        ss_lengths: an array of single-byte values specifying the lengths of search results (from the end) to be hashed
+             in ascending order. For example, if *ss_lengths==[1, 2]*, *ss_search=="aC" and *cs==False*, the searched strings 
+             hashed for "spaCy" would be "c" and "ca".
+    
+        Many of the buffers passed into and used by this method contain single-byte numerical values. This takes advantage of 
+        the fact that we are hashing short affixes and searching for small groups of characters. The calling code is responsible
+        for ensuring that lengths being passed in cannot exceed 63 and hence that resulting values with maximally four-byte 
+        character widths can never exceed 255.
+
+        Note that this method performs no data validation itself as it expects the calling code will already have done so, and
+        that the behaviour of the code may be erratic if the supplied parameters do not conform to expectations.
+        """
+        
+        # Work out lengths
+        cdef int p_lengths_l = strlen(<char*> p_lengths)
+        cdef int s_lengths_l = strlen(<char*> s_lengths)
+        cdef int ps_lengths_l = strlen(<char*> ps_lengths)
+        cdef int ss_lengths_l = strlen(<char*> ss_lengths)
+        cdef int hashes_per_tok = p_lengths_l + s_lengths_l + ps_lengths_l + ss_lengths_l
+        cdef int p_max_l = p_lengths[p_lengths_l - 1] if p_lengths_l > 0 else 0
+        cdef int s_max_l = s_lengths[s_lengths_l - 1] if s_lengths_l > 0 else 0
+        cdef int ps_max_l = ps_lengths[ps_lengths_l - 1] if ps_lengths_l > 0 else 0
+        cdef int ss_max_l = ss_lengths[ss_lengths_l - 1] if ss_lengths_l > 0 else 0
+        
+        # Define / allocate buffers
+        cdef Pool mem = Pool()
+        cdef unsigned char* pref_l_buf = <unsigned char*> mem.alloc(p_max_l, 1)
```

Comment:
I'd use `std::vector` for all of these, they give you the determenistic destruction. (Or perhaps `std::string` which is just a container for raw byte data.)

(Even though `char` is always 1 byte on modern platforms, I'd use `sizeof` rather than raw sizes, conveys the semantics better.)

Reply:
I looked into replacing these with C++ types, but came across the problem (presuming that it makes sense to use the same type for all buffers, which stylistically I would say is the case) that:

- the `_write_hashes()` method has to be able to accept both the raw token string returned from the `StringStore` and the contents of one of these buffers as values of the same variable. This essentially precludes using `std::vector`.
- it would be possible to change the new `StringStore` method to return `std::string` rather than `char*`, but if we want to use `std::string` throughout we have the problem of how to allocate the buffers in the code marked here. While with `char*` we can simply allocate the memory, retaining whatever already happens to be there because the relevant part of it will be overwritten anyway, it seems that with `std:string` we have to initialize the memory with something.

Doesn't `cymem.Pool` give you effective memory management as well?

Reply:
I am not sure what gcc/C++ version we have access to in the wheel builds, but newer C++ versions have `string_view`, which can be used to provide views for both borrowed strings (like strings from the string strore) and owned strings (`std::string`). This can be used to provide a generic API for functions that need to deal with owned and borrowed strings.

Reply:
`string_view` isn't available via `libcpp`, so using it would require wrapping the relevant library from C++17. I'm not sure whether the additional safety is worth the considerably more complex program architecture (which would be a potential source of bugs in itself). I'd say no, but feel free to go ahead and make the change if you think it makes sense!